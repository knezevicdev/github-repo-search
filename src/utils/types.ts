export type RepositoryType = 'public' | 'private' | 'fork';
export type RepositoryTypeFilter = 'all' | RepositoryType;

export interface Repository {
  id: number;
  private: boolean;
  fork: boolean;
  name: string;
  html_url: string;
}
