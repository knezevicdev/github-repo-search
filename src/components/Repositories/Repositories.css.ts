import styled from 'styled-components';
import colors from 'styled/colors';

export const Repository = styled.li`
  padding: 10px;
  border-bottom: 1px solid ${colors.gray1};
  display: flex;
  align-items: center;
`;
