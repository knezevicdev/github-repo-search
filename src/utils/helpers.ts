import { Repository, RepositoryType } from 'utils/types';

export const getRepositoryType = (repository: Repository): RepositoryType => {
  if (repository.fork) {
    return 'fork';
  }

  if (repository.private) {
    return 'private';
  }

  return 'public';
};
