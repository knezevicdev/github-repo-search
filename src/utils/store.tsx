import React, { ComponentType } from 'react';
import { observable, computed, comparer, action } from 'mobx';

import { RepositoryTypeFilter, Repository } from 'utils/types';
import { getRepositoryType } from 'utils/helpers';

const storeContext = React.createContext<Store | null>(null);

class Store {
  @observable isLoading = true;
  @observable hasError = false;
  @observable private allRepositories: Repository[] = [];
  @observable searchText = '';
  @observable typeFitler: RepositoryTypeFilter = 'all';
  @observable showingAll = false;

  @computed({ equals: comparer.shallow })
  private get filteredRepositories(): Repository[] {
    const searchText = this.searchText.trim().toLowerCase();
    let repositories: Repository[] = this.allRepositories;

    if (this.typeFitler !== 'all' || searchText !== '') {
      repositories = repositories.filter((repository) => {
        if (this.typeFitler !== 'all' && this.typeFitler !== getRepositoryType(repository)) return false;
        if (repository.name.toLowerCase().indexOf(searchText) === -1) return false;

        return true;
      });

      if (searchText !== '') {
        const reg = new RegExp(searchText, 'gi');

        repositories = repositories.map((repository: Repository) => ({
          ...repository,
          name: repository.name.replace(reg, (str: string) => `<em>${str}</em>`),
        }));
      }
    }

    return repositories;
  }

  @computed({ equals: comparer.shallow })
  get repositories() {
    if (!this.showingAll) return this.filteredRepositories.slice(0, 3);

    return this.filteredRepositories;
  }

  @computed
  get repositoriesCount() {
    return this.filteredRepositories.length;
  }

  @action.bound
  async fetchRepositories(githubUser: string) {
    try {
      const repositories = await fetch(`https://api.github.com/users/${githubUser}/repos`);
      this.allRepositories = await repositories.json();
    } catch {
      this.hasError = true;
    }

    this.isLoading = false;
  }

  @action.bound
  setTypeFilter(type: RepositoryTypeFilter) {
    this.typeFitler = type;
  }

  @action.bound
  setSearchText(searchText: string) {
    this.searchText = searchText;
  }

  @action.bound
  toggleShowingAll() {
    this.showingAll = !this.showingAll;
  }
}

export const StoreProvider: React.FC = ({ children }) => {
  const store = new Store();

  return <storeContext.Provider value={store}>{children}</storeContext.Provider>;
};

export const withStore = (WrappedComponent: ComponentType): React.FC => {
  const StoreHOC: React.FC = (props) => {
    return (
      <StoreProvider>
        <WrappedComponent {...props} />
      </StoreProvider>
    );
  };

  return StoreHOC;
};

export const useStore = (): Store => {
  const store = React.useContext(storeContext);
  if (!store) {
    throw new Error('useStore must be used within a StoreProvider.');
  }
  return store;
};
