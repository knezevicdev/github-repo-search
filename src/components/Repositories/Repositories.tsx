import React from 'react';

import { observer } from 'mobx-react';
import { useStore } from 'utils/store';

import { Link } from 'styled/components';

import { Icon } from 'components';

import { Repository } from './Repositories.css';
import { getRepositoryType } from 'utils/helpers';

const Repositories: React.FC = () => {
  const { repositories } = useStore();

  return (
    <ul>
      {repositories.map((repository) => (
        <Repository key={`repository#${repository.id}`}>
          <Icon type={getRepositoryType(repository)} />
          &nbsp;
          <Link href={repository.html_url} dangerouslySetInnerHTML={{ __html: repository.name }} />
        </Repository>
      ))}
    </ul>
  );
};

export default observer(Repositories);
