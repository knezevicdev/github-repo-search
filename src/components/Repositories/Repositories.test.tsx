import React from 'react';
import { shallow } from 'enzyme';

import { StoreProvider } from 'utils/store';

import Repositories from './Repositories';

describe('Repositories', () => {
  it('renders snapshots', () => {
    const wrapper = shallow(
      <StoreProvider>
        <Repositories />
      </StoreProvider>,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
