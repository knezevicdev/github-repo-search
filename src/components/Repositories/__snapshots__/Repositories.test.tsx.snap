// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repositories renders snapshots 1`] = `
<ContextProvider
  value={
    Store {
      "allRepositories": Array [],
      "hasError": false,
      "isLoading": true,
      "searchText": "",
      "showingAll": false,
      "typeFitler": "all",
    }
  }
>
  <Repositories />
</ContextProvider>
`;
